### PR TITLE
Add Feature - Support overlapping all-to-all with FFN computation in MoE layer

### PR DESCRIPTION
**Pull request by yzygitzh**
This is a PoC to overlap computation and communication of A2A-FFN-A2A phase in MoE layer.

To achieve this, data used in A2A-FFN-A2A phase is split into multiple chunks along batch size dimension. As a result, each chunk can be processed independently in both forward and backward pass.

### CODE REVIEW

File: tutel/impls/moe_layer.py [hunk 1]
```diff
@@ -134,13 +134,22 @@ def apply_on_expert_fn(self, input, expert_fn, group, sharded_count):
...
+
+        # Implicit x.contiguous() in CurrentStreamRelease.forward() and CurrentStreamAcquire.backward()
+        dispatched_input_ready = CurrentStreamRelease.apply(dispatched_input, 0)
+        dispatched_input_scattered_after_a2a = AllToAllScatterAsync.apply(dispatched_input_ready)
+
+        expert_output_scattered = [
+            CurrentStreamRelease.apply(expert_fn(CurrentStreamAcquire.apply(x, i)).to(input.dtype), i)
+            for i, x in enumerate(dispatched_input_scattered_after_a2a)]
+        expert_output_gathered_after_a2a = AllToAllGatherAsync.apply(*expert_output_scattered)
+        expert_output = CurrentStreamAcquire.apply(expert_output_gathered_after_a2a, 0)
```
Comment (ghostplant) [inline tutel/impls/moe_layer.py:152]: Is it possible to set an option to enable this instead of directly replacing the original approach?

Comment (yzygitzh) [reply tutel/impls/moe_layer.py:152]: Sure, will make this an option in formal PR later.

File: tutel/impls/moe_layer.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def __init__(
         self,
         model_dim,
         num_global_experts,
+        a2a_ffn_overlap_degree,
         capacity_factor=1.0,
```
Comment (ghostplant) [inline tutel/impls/moe_layer.py:50]: Better to change `a2a_ffn_overlap_degree` to `a2a_ffn_overlap_degree=1` ?

File: tutel/impls/moe_layer.py [hunk 3]
```diff
@@ -73,6 +74,8 @@ def __init__(
         input_dropout_p = kwargs.get('input_dropout_p', 0)
         self.input_dropout = torch.nn.Dropout(p=input_dropout_p) if input_dropout_p else None
 
+        self.a2a_ffn_overlap_degree = a2a_ffn_overlap_degree
+
```
Comment (ghostplant) [inline tutel/impls/moe_layer.py:78]: In order not to change the original usage of TopKGate, better to set `a2a_ffn_overlap_degree=1` by default or using `kwargs.get('a2a_ffn_overlap_degree', 1)` to access the value.

Comment (yzygitzh) [reply tutel/impls/moe_layer.py:78]: Fixed.

**Pull request merged**